src/plugins/discover/public/application/main/hooks/use_adhoc_data_views.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { DataViewsContract, type DataView } from '@kbn/data-views-plugin/public';
+import { SavedSearch } from '@kbn/saved-search-plugin/public';
+import {
+  UPDATE_FILTER_REFERENCES_ACTION,
+  UPDATE_FILTER_REFERENCES_TRIGGER,
+} from '@kbn/unified-search-plugin/public';
+import { ActionExecutionContext } from '@kbn/ui-actions-plugin/public';
+import { useConfirmPersistencePrompt } from '../../../hooks/use_confirm_persistence_prompt';
+import { getUiActions, getUrlTracker } from '../../../kibana_services';
+import { GetStateReturn } from '../services/discover_state';
+
+export const useAdHocDataViews = ({
+  dataView,
+  savedSearch,
+  dataViews,
+  stateContainer,
+  onChangeDataView,
+}: {
+  dataView: DataView;
+  savedSearch: SavedSearch;
+  dataViews: DataViewsContract;
+  stateContainer: GetStateReturn;
+  onChangeDataView: (dataViewId: string) => Promise<void>;
+}) => {
+  const [adHocDataViewList, setAdHocDataViewList] = useState<DataView[]>(
+    !dataView.isPersisted() ? [dataView] : []
+  );
+
+  useEffect(() => {
+    if (!dataView.isPersisted()) {
+      setAdHocDataViewList((prev) => {
+        const existing = prev.find((prevDataView) => prevDataView.id === dataView.id);
+        return existing ? prev : [...prev, dataView];
+      });
+    }
+  }, [dataView]);
+
+  /**
+   * When saving a saved search with an ad hoc data view, a new id needs to be generated for the data view
+   * This is to prevent duplicate ids messing with our system
+   */
+  const updateAdHocDataViewId = useCallback(
+    async (dataViewToUpdate: DataView) => {
+      const newDataView = await dataViews.create({ ...dataViewToUpdate.toSpec(), id: undefined });
+
+      dataViews.clearInstanceCache(dataViewToUpdate.id);
```

Comment:
should we really do this ? possibly the instance was created in lens, now i go back to lens and its gone ?

Reply:
If you saved the Lens vis with the ad-hoc data view, Lens will re-create it. If you didn't save it, you can't go back anyway. I think this behavior is correct. When navigating between apps, the spec should be passed along with the navigation, so the target app can re-create it. This is important as we can't rely on the cache on cross-app navigations anyway (we don't know what happens in between, might even be a full page refresh)

Reply:
what if user clicked back in the browser, or just navigated to lens via the "recently viewed" links ?

Reply:
As Lens doesn’t store the current state in the url (and discard it once the user navigates away from the app), you would get an empty editor without ad-hoc data view. This is the case in general

Reply:
`Explore in Discover` anyway opens new browser tab. So, then doing `clearInstanceCache` in different tab will be safe. Lens -> Discover navigation with adhoc data views will be improved in follow up, after merging current PR.

Reply:
The only thing here is when we create adhoc data view from Discover, then navigating to Lens. On modifing data view in Lens, `clearInstanceCache` also will be called. Then, trying to click history back button, will give a toast warn about missing data view in Discover.

Reply:
@flash1293 is it intended that way?